### ISSUE TITLE
Request: A tweak to Auto Tagging

### ISSUE THREAD

**Issue by NaPurHab** [BiglySoftware/BiglyBT#1233]
Auto Tagging is working great. Just one small tweak is needed.  When the Open Torrent dialog comes up, the tags are working as they should. However sometimes I want to change the Tags.  Currently if say the Tag is VIDEO it is LOCKED.  I run a PLEX server and sometimes I am placing the VIDEO into PLEX directly. Where normally it goes to VIDEO and I Transcode it to the format I want on PLEX.

So the tweak I would like is that AutoTagging sets the TAG, but it is not locked but changeable in the OPEN TORRENTS Dialog.

**Comment by newgithubuser2019** [comment]
Second this

**Comment by parg** [comment]
There will be a new option under Interface->Tags to allow this to happen in B39

**Comment by ferdnyc** [comment]
+1 — and I'm curious how exactly the B39 option will work.

It's always been a minor frustration of mine that, if I have auto-tags configured based on matching rules (like, the ones which auto-assign based on tracker hostname, or etc.), then I **can't** manually apply or unapply those tags. It's led to me effectively having duplicate tags, since I ended up creating duplicates of the auto-tags with the same properties, but without the auto-apply rules, just so I could  assign them manually when needed.

If I'm understanding the request correctly, eliminating that distinction between auto and manual tags, and allowing tags with auto-apply rules to _also_ be applied/removed manually (as an "override" of the tag's auto-apply rules, I guess), would not only solve this, it would be useful for all sorts of other scenarios as well.

**Comment by parg** [comment]
Ah, they're talking about the open-torrent-option dialog - there are some separate auto-tagging rules under Tools->Options->Interface->Tags that are applied just in the open-torrent-options window when selecting the initial tags, nothing to do with the real tags themselves.

By the way, there is also a normal Tag constraint scope of 'new downloads' that allows you to specify a constraint that will only be applied once when a download is added.

**Comment by ferdnyc** [comment]
> By the way, there is also a normal Tag constraint scope of 'new downloads' that allows you to specify a constraint that will only be applied once when a download is added.

Ooh, neat, I think that's newer than the last time I looked.. ...Does that allow the tag to be manually applied, as well? 

Actually, looking at my setup again, I guess I mostly worked around the auto-tagging limitation after I bumped into it, by either having the tag assigned by RSSFeed Scanner, or by the "Assign to tag:" option of Files / Torrents — both of which count as "manual" tagging.

Only two of my tags are real auto-tags. One matches on the tracker hostnames for a certain private tracker, and sets some rules. The second uses a `hasTag()` constraint for the _first_ tag, to apply some **additional** rules.

I split things up that way, I remember now, because set up the way I have it, I **can** manually apply or remove the _second_ tag. The first one, with the "Trackers" property set on it, seems completely inaccessible to me for manual use.

* If it's not set on a torrent, then it doesn't even appear in the Tags section of the context menu.
* If it _is_ set on a torrent, then trying to remove it from the context menu has no effect. (Maybe that's because I need to switch the scope to "New downloads", though?)

So, the rules from the second tag (the `hasTag()` constraint one) _could_ be part of the first tag (the tracker-matching one), except I wanted the ability to unapply them in certain situations. So I had to set up the cascaded double-tagging setup I use, to be able to do that.

**Comment by parg** [comment]
Cascading tags is a good way to go.

The 'tracker properties' approach to automation came along before Tag constraints. I suppose constraints could be extended to include a 'hasTracker(...)' function, thus replacing the 'tracker properties' but for the moment the best way to combine the two is via your approach.

With the new-downloads constraint scope you can manually assign/remove from that tag, the automation is applied once and only once on download addition and then effectively forgotten for that download.

**Comment by NaPurHab** [comment]
> There will be a new option under Interface->Tags to allow this to happen in B39

Why does it need to be an option?  Just like when I set the IGNORE options, I can change the selected files in the Open Torrents Dialog.  The Auto-Tagging should be just the same, set the state of the tag when opening the dialog. But allow the user to change before clicking OK.  No need for a NEW Option, just Unlock it.

**Comment by parg** [comment]
because auto-tagging works on the currently selected files - if you change the file selection for the torrent then the auto-tagging adjusts accordingly

**Comment by NaPurHab** [comment]
Only other change would be to allow something like:

60.minutes*.mkv;panorama*.mkv          Tag=PLEX

Then add a way to reorder them and an Option to set only the first matching.  For Example the tag VIDEO would have *.mkv in it. So I would want PLEX before VIDEO and if PLEX matched first, don't set VIDEO.

Perhaps Change the Add another tag BUTTON, to a Label. Then have 2 Buttons after labeled BEFORE and AFTER.  I'll leave it up to you.

**Comment by NaPurHab** [comment]
> because auto-tagging works on the currently selected files - if you change the file selection for the torrent then the auto-tagging adjusts accordingly

Ok, I assumed it would be set the state of the TAG to ON when Opening the Dialog, but not lock it. Just like the Ignore sets selected files OFF. But then I can change the selected files before closing the dialog.

**Comment by parg** [comment]
If you want complex auto-tagging then don't do it in the open-torrent-options dialog, set up tags with the required tag constraints and scope of 'new download'

**Comment by ferdnyc** [comment]
> The 'tracker properties' approach to automation came along before Tag constraints. I suppose constraints could be extended to include a 'hasTracker(...)' function, thus replacing the 'tracker properties' 

That sounds sensible to me. (And I imagine would make the code a bit cleaner, too, to have tracker matching exist as just another type of constraint, rather than this whole other tag dimension.) But it also doesn't strike me as any sort of priority. As you say, the cascading works just fine for now.
